index.bs

```diff
@@ -94,10 +109,60 @@ Note: For an [=endpoint node=] the hostname in the above steps will be
 typically be "<code>localhost</code>".
 
 <div algorithm>
-To <dfn>handle an incoming message</dfn> given a
-[=WebSocket connection=] <var ignore>connection</var>:
+To <dfn>handle an incoming message</dfn> given a [=WebSocket connection=]
+|connection|, type |type| and data |data|:
+
+ 1. If |type| is not [=%x1 denotes a text frame|text=], return.
+
+    Issue: Should we instead close |connection| with [=status
+    codes|status code=] 1003, or [=respond with an error=]?
+
+ 2. [=Assert=]: |data| is a [=scalar value string=], because the
+     WebSocket [=handling errors in UTF-8-encoded data=] would already
+     have [=fail the WebSocket connection|failed the WebSocket
+     connection=] otherwise.
+
+    Issue: Nothing seems to define what [=status codes|status code=]
+    is used for UTF-8 errors.
+
+ 3. Let |parsed| be the result of [=parse JSON into Infra
+    values|parsing JSON into Infra values=] given |data|. If this
+    throws an exception, then [=respond with an error=] given
+    |connection| and [=error code=] [=invalid argument=], and finally
+    return.
+    <!-- corresponds to Parse error (-32700) in JSON-RPC -->
+
+ 4. If any of the following conditions are false:
+
+     1. |parsed| is a [=map=]
+
+     2. |parsed|["<code>id</code>"] <a for=map>exists</a> and is an
+        integer in the range [−2147483648, 2147483647].
+
+     3. |parsed|["<code>method</code>"] <a for=map>exists</a> and is a
+        string.
+
+     4. |parsed|["<code>params</code>"], if it <a for=map>exists</a>,
+        is a [=map=].
+
+     Issue: Require that there are no other keys?
```

Comment:
Yep, for parameters that certainly makes sense. I noticed while experimenting with CDP that it fails if there are any additional things at this layer. Worth raising with potential implementers and others I think. I'll update the description a bit.

Reply:
If we ever wanted to add a version field or something it might cause problems in old clients. But I suppose in that case we could negotiate the version in the capabilities and always send v1 messages to a v1 client.

So I think I agree that having a fixed set of fields at this level would be OK. But compared to devtools we have much less ability to make unilateral changes to the protocol, so the tradeoffs might in general be different.

Reply:
Are you happy with the issue being left in for discussion? I think it's worth highlighting minor points of doubt and removing them when we've decided, explicitly or not. (Unlike TODOs in code, these are red and terrible, I won't be tempted to leave them around for a long time.)

Reply:
Yes.